sycl/include/sycl/esimd/detail/esimd_host_util.hpp

```diff
@@ -0,0 +1,362 @@
+template <> struct restype_ex<float, long long> { typedef float type; };
+template <> struct restype_ex<float, unsigned long long> {
+  typedef float type;
+};
+
+template <> struct restype_ex<double, char> { typedef double type; };
+template <> struct restype_ex<double, unsigned char> { typedef double type; };
+template <> struct restype_ex<double, short> { typedef double type; };
+template <> struct restype_ex<double, unsigned short> { typedef double type; };
+template <> struct restype_ex<double, int> { typedef double type; };
+template <> struct restype_ex<double, unsigned int> { typedef double type; };
+template <> struct restype_ex<double, half> { typedef double type; };
+template <> struct restype_ex<double, float> { typedef double type; };
+template <> struct restype_ex<double, double> { typedef double type; };
+template <> struct restype_ex<double, long long> { typedef double type; };
+template <> struct restype_ex<double, unsigned long long> {
+  typedef double type;
+};
+
+template <> struct restype_ex<long long, char> { typedef long long type; };
+template <> struct restype_ex<long long, unsigned char> {
+  typedef long long type;
+};
+template <> struct restype_ex<long long, short> { typedef long long type; };
+template <> struct restype_ex<long long, unsigned short> {
+  typedef long long type;
+};
+template <> struct restype_ex<long long, int> { typedef long long type; };
+template <> struct restype_ex<long long, unsigned int> {
+  typedef long long type;
+};
+template <> struct restype_ex<long long, half> { typedef half type; };
+template <> struct restype_ex<long long, float> { typedef float type; };
+template <> struct restype_ex<long long, double> { typedef double type; };
+template <> struct restype_ex<long long, long long> { typedef long long type; };
+template <> struct restype_ex<long long, unsigned long long> {
+  typedef long long type;
+};
+
+template <> struct restype_ex<unsigned long long, char> {
+  typedef long long type;
+};
+template <> struct restype_ex<unsigned long long, unsigned char> {
+  typedef long long type;
+};
+template <> struct restype_ex<unsigned long long, short> {
+  typedef long long type;
+};
+template <> struct restype_ex<unsigned long long, unsigned short> {
+  typedef long long type;
+};
+template <> struct restype_ex<unsigned long long, int> {
+  typedef long long type;
+};
+template <> struct restype_ex<unsigned long long, unsigned int> {
+  typedef long long type;
+};
+template <> struct restype_ex<unsigned long long, half> { typedef half type; };
+template <> struct restype_ex<unsigned long long, float> {
+  typedef float type;
+};
+template <> struct restype_ex<unsigned long long, double> {
+  typedef double type;
+};
+template <> struct restype_ex<unsigned long long, long long> {
+  typedef long long type;
+};
+template <> struct restype_ex<unsigned long long, unsigned long long> {
+  typedef long long type;
+};
+
+template <typename T> struct abstype;
```

Comment:
Also do not use `typedef` in modern code...

Reply:
Modern or Not Modern.  This is perfect legal C++ code.
"typedef T type" or "using type = T" achieve the same thing. I prefer to keep it as it is

Reply:
> Modern or Not Modern. This is perfect legal C++ code.
> "typedef T type" or "using type = T" achieve the same thing. I prefer to keep it as it is

@cmc-rep, could you clarify why do you prefer `typedef` over `using`?

We applied [recommendations from authors of C++](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rt-using) to our project and it would be nice to the code base consistent.

Reply:
I have no preference. This code is ported from the existing CM code. There are a lot of typedefs in this file. If I have to change it manually, it is going to be quite laborous work. If someone has a script that can do that, I am happy to apply it.

Reply:
There is a clang-tidy tool which can do this transformation.
https://clang.llvm.org/extra/clang-tidy/checks/modernize-use-using.html#modernize-use-using

http://clang.llvm.org/extra/clang-tidy/index.html#using-clang-tidy

If you are able to apply this tool to the patch, it will resolve the comment from Ronan.

Reply:
fixed with clang-tidy

Reply:
Resolved